src/java.desktop/macosx/native/libawt_lwawt/awt/a11y/CommonComponentAccessibility.m

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#import "CommonComponentAccessibility.h"
+#import "JNIUtilities.h"
+#import "ThreadUtilities.h"
+
+static jclass sjc_CAccessibility = NULL;
+static jmethodID sjm_getAccessibleComponent = NULL;
+
+#define GET_ACCESSIBLECOMPONENT_STATIC_METHOD_RETURN(ret) \
+    GET_CACCESSIBILITY_CLASS_RETURN(ret); \
+    GET_STATIC_METHOD_RETURN(sjm_getAccessibleComponent, sjc_CAccessibility, "getAccessibleComponent", \
+           "(Ljavax/accessibility/Accessible;Ljava/awt/Component;)Ljavax/accessibility/AccessibleComponent;", ret);
+
+static NSMutableDictionary * _Nullable rolesMap;
+
+/*
+ * Common ancestor for all the accessibility peers that implements the new method-based accessibility API
+ */
+@implementation CommonComponentAccessibility
+
++ (void) initializeRolesMap {
+    /*
+     * Here we should keep all the mapping between the accessibility roles and implementing classes
+     */
+    rolesMap = [[NSMutableDictionary alloc] initWithCapacity:1];
+
+    [rolesMap setObject:@"ButtonAccessibility" forKey:@"pushbutton"];
+}
+
+/*
+ * If new implementation of the accessible component peer for the given role exists
+ * return the allocated class otherwise return nil to let old implementation being initialized
+ */
++ (JavaComponentAccessibility *) getComponentAccessibility:(NSString *)role
+{
+    if (rolesMap == nil) {
+        [self initializeRolesMap];
```

Comment:
Is it possible that this will be called on the different threads? Or it is always appkit(probably AWT_ASSERT_APPKIT_THREAD could be useful)?

Reply:
Well, this method is being called from AppKit thread only but giving that it is static and public adding an assertion would be indeed wise. Fixed.